File: src/main/java/example/signup/SignupSheet.java
package example.signup;

import java.util.LinkedHashSet;
import java.util.Set;

public class SignupSheet {
    private final SessionId sessionId;
    private final int capacity;
    private final LinkedHashSet<AttendeeId> signups = new LinkedHashSet<>();
    private boolean isClosed = false;

    public SignupSheet(SessionId sessionId, int capacity) {
        this.sessionId = sessionId;
        this.capacity = capacity;
    }

    public SessionId getSessionId() {
        return sessionId;
    }

    public int getCapacity() {
        return capacity;
    }

    public boolean isFull() {
        return signups.size() == capacity;
    }

    public boolean isClosed() {
        return isClosed;
    }

    public Set<AttendeeId> getSignups() {
        return Set.copyOf(signups);
    }

    public boolean isSignedUp(AttendeeId attendeeId) {
        return signups.contains(attendeeId);
    }

    public void signUp(AttendeeId attendeeId) {
        if (isClosed() || isFull()) throw new IllegalStateException();
        signups.add(attendeeId);
    }

    public void cancelSignUp(AttendeeId attendeeId) {
        if (isClosed()) throw new IllegalStateException();

        signups.remove(attendeeId);
    }

    public void close() {
        isClosed = true;
    }
}

File: src/main/java/example/signup/SignupSheet.kt
package example.signup


sealed class SignupSheet {
    abstract val sessionId: SessionId
    abstract val capacity: Int
    abstract val signups: Set<AttendeeId>
    
    fun isSignedUp(attendeeId: AttendeeId): Boolean {
        return attendeeId in signups
    }
}

sealed class Open : SignupSheet() {
    fun cancelSignUp(attendeeId: AttendeeId): SignupSheet {
        return Available(sessionId, capacity, signups = signups - attendeeId)
    }
    
    fun close(): SignupSheet {
        return Closed(sessionId, capacity, signups)
    }
}

data class Available @JvmOverloads constructor (
    override val sessionId: SessionId,
    override val capacity: Int,
    override val signups : Set<AttendeeId> = emptySet(),
) : Open() {
    
    fun signUp(attendeeId: AttendeeId): SignupSheet {
        val newSignups = signups + attendeeId
        return when (newSignups.size) {
            capacity -> Full(sessionId, newSignups)
            else -> copy(signups = newSignups)
        }
    }
}

data class Full(
    override val sessionId: SessionId,
    override val signups: Set<AttendeeId>
) : Open() {
    override val capacity: Int = signups.size
}

data class Closed(
    override val sessionId: SessionId,
    override val capacity: Int,
    override val signups: Set<AttendeeId>
) : SignupSheet() {
}